src/null-safety/migration-guide.md

```diff
@@ -1,23 +1,453 @@
+var zero = ints[0];
+var one = zero + 1;
+var zeroOne = <int>[zero, one];
+```
+
+The default migration is backward compatible but not ideal:
+
+```dart
+var ints = const <int?>[0, null];
+var zero = ints[0];
+var one = zero! + 1;
+var zeroOne = <int?>[zero, one];
+```
+
+By clicking line 3, you can see the migration tool's reasons for
+adding the `!`.
+But if you know that `zero` can't be null,
+then you can improve the migration result.
+
+
+#### Improving migration results
+
+When analysis infers the wrong nullability,
+you can override its proposed edits by inserting temporary
+`/*?*/` and `/*!*/` markers:
+
+* In the **Edit Details** pane of the migration tool,
+  you can insert these markers using the
+  **Add `/*?*/` hint** and **Add `/*!*/` hint** buttons.
+
+  These buttons add comments to your file immediately,
+  and there's **no Undo**.
+  If you don't want a hint that the tool inserted,
+  you can use your usual code editor to remove it.
+
+* You can use an editor to add these markers,
+  even while the tool is still running.
+  Because your code hasn't opted into null safety yet,
+  you can't use new null-safety features.
+  You can, however, make changes like refactoring
+  that don't depend on null-safety features.
+
+  When you've finished editing your code,
+  click **Rerun from sources** to pick up your changes.
+
+A single hint can have ripple effects elsewhere in the code.
+In the example from before,
+adding a `/*!*/` marker where `zero` is assigned its value (on line 2)
+makes the migration tool infer the type of `zero` as `int` instead of `int?`.
+This type change can affect code that directly or indirectly uses `zero`.
+
+```dart
+var zero = ints[0]/*!*/;
+```
+
+With the above hint, the migration tool changes its proposed edits,
+as the following code snippets show.
+Line 3 no longer has a `!` after `zero`,
+and in line 4 `zeroOne` is inferred to be
+a list of `int`, not `int?`.
+
+<table markdown="1">
+<tr>
+<th>First migration</th>
+<th>Migration with hint</th>
+</tr>
+<tr markdown="1">
+  <td markdown="1">
+```dart
+var ints = const <int?>[0, null];
+var zero = ints[0];
+var one = zero! + 1;
+var zeroOne = <int?>[zero, one];
+```
+  </td>
+  <td markdown="1">
+```dart
+var ints = const <int?>[0, null];
+var zero = ints[0]/*!*/;
+var one = zero + 1;
+var zeroOne = <int>[zero, one];
+```
+  </td>
+</tr>
+</table>
+
+#### Applying changes
+
+When you like all of the changes
+that the migration tool proposes, click
+**Apply migration** to tell the tool to save the migrated code
+and to update the minimum SDK in your pubspec.
+
+The next step is to [statically analyze your code](#step3-analyze).
+If it's valid, then [test your code](#step4-test).
+Then, if you've published your code on pub.dev,
+[publish a null-safe prerelease](#step5-publish).
+
+
+### Migrating by hand
+
+If you prefer not to use the migration tool,
+you can migrate manually.
+Because Dart supports [mixed-version programs][Unsound null safety],
```

Comment:
This isn't consistent with the workflow below.  It's true you can do this, but not if you follow the workflow below - in order to migrate one library at a time, you have to opt all of the other libraries out explicitly by marking them with `// @dart = 2.9`.  @mit-mit @devoncarew do we really want to go into that here?

Reply:
I don't think so. I think the present page should focus on sound migration (that is what we preached in Step 1. above), and then cover how to do a sound migration with either the migration tool or the analyzer.

We have a dedicated topic for unsound underway (https://github.com/dart-lang/site-www/pull/2764) so we can cover all the gory details of that here.

Concretely, on the current page, I suggest we remove "Because Dart supports mixed-version programs, you can migrate one library (generally one Dart file) at a time, while still being able to run your app and its tests."